community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/ValuesExtension.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.impl.schema;
+
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.lifecycle.LifeSupport;
+import org.neo4j.kernel.lifecycle.Lifecycle;
+import org.neo4j.values.storable.Values;
+
+public class ValuesExtension implements Lifecycle
+{
+    private final LifeSupport life = new LifeSupport();
+    private Config configuration;
+
+    ValuesExtension( ValuesKernelExtensionFactory.Dependencies dependencies )
+    {
+        configuration = dependencies.getConfig();
+    }
+
+    @Override
+    public void init()
+    {
+        life.init();
+    }
+
+    @Override
+    public void start()
+    {
+        life.start();
+        int nanoPrecisionConfig = configuration.get( GraphDatabaseSettings.temporal_nanosecond_precision );
+
+        // If two db instances in the same JVM try to set this to different values, the last setting will override the previous one
+        if ( nanoPrecisionConfig != Values.getNanoPrecision() && Values.nanoPrecisionConfigAltered )
+        {
+            throw new IllegalArgumentException( "dbms.temporal.nanosecond_precision can only be set once" );
+        }
+        Values.setNanoPrecision( nanoPrecisionConfig );
+        Values.nanoPrecisionConfigAltered = true;
+    }
+
+    @Override
+    public void stop()
+    {
+        life.stop();
+    }
+
+    @Override
+    public void shutdown()
+    {
+        Values.nanoPrecisionConfigAltered = false;
```

Comment:
Why set it to false here? Then if we start up three databases, and then shut down one of them it will set this to false, so starting a fourth database could change the setting of the first two still running. I think the only way for a user to change this setting should be JVM restart. That matches current settings behaviour for most neo4j.conf settings anyway.

Reply:
On option is to make `nanoPrecisionConfigAltered` private, so it can only be set within `Values.setNanoPrecision` and then it will also only ever change from false to true and never back again.